examples/kgcn/KGCN.py

```diff
@@ -0,0 +1,136 @@
+import numpy as np
+import argparse
+from model import KGCN
+from data_loader import DataLoader
+import torch
+import torch.optim as optim
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import roc_auc_score
+from gs.utils import SeedGenerator
+from tqdm import tqdm
+
+
+# Dataset class
+class KGCNDataset(torch.utils.data.Dataset):
+    def __init__(self, df):
+        self.df = df
+
+    def __len__(self):
+        return len(self.df)
+
+    def __getitem__(self, idx):
+        user_id = np.array(self.df['userID'][idx])
+        item_id = np.array(self.df['itemID'][idx])
+        label = np.array(self.df['label'][idx], dtype=np.float32)
+        return user_id, item_id, label
+
+
+def train(args):
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    print('device: ', device)
+
+    # build dataset and knowledge graph
+    data_loader = DataLoader(args.dataset)
+    kg = data_loader.load_kg()
+    df_dataset = data_loader.load_dataset()
+
+    # train test split
+    x_train, x_test, y_train, y_test = train_test_split(
+        df_dataset, df_dataset['label'], test_size=1 - args.ratio, shuffle=False, random_state=999)
+    userID = torch.tensor(df_dataset['userID']).to(device)
+    itemID = torch.tensor(df_dataset['itemID']).to(device)
+    labels = torch.tensor(df_dataset['label'], dtype=torch.float32).to(device)
+    # x_test.reset_index(inplace=True, drop=True)
+    # train_dataset = KGCNDataset(x_train)
+    # test_dataset = KGCNDataset(x_test)
+    train_seedloader = SeedGenerator(
+        torch.tensor(x_train.index), batch_size=args.batch_size, shuffle=True, drop_last=False)
+    test_seedloader = SeedGenerator(
+        torch.tensor(x_test.index), batch_size=args.batch_size, shuffle=True, drop_last=False)
+    # train_loader = torch.utils.data.DataLoader(
+    #     train_dataset, batch_size=args.batch_size)
+    # test_loader = torch.utils.data.DataLoader(
+    #     test_dataset, batch_size=args.batch_size)
+
+    # prepare network, loss function, optimizer
+    num_user, num_entity, num_relation = data_loader.get_num()
+    user_encoder, entity_encoder, relation_encoder = data_loader.get_encoders()
+    net = KGCN(num_user, num_entity, num_relation, kg, args, device).to(device)
+    criterion = torch.nn.BCELoss()
+    optimizer = optim.Adam(net.parameters(), lr=args.lr,
+                           weight_decay=args.l2_weight)
+
+    # train
+    loss_list = []
+    test_loss_list = []
+    auc_score_list = []
+
+    for epoch in range(args.n_epochs):
+        running_loss = 0.0
+        for i, seeds in enumerate(tqdm(train_seedloader)):
```

Comment:
Can we write a version that uses online sampling to train the KGCN model? Online sampling means that the subgraph is generated at each iteration.

Reply:
Both the author's version (implemented with tensorflow) and the pytorch alternative implement KGCN by pre-sampling on the whole graph. And I think we need to align with them.

Reply:
This version is ok. But an online sampling verison may be more attractive and it's not difficult for ours to support it. In fact, the early graphsage was also an offline sampling implementation. However, online sampling's graphsage is more popular.

Reply:
Yep, it's just a minor convertion and I will mark it as `@TODO`. BTW, We may still need to keep this current version in the experiments for alignment.

Reply:
Altered to online sampling.